Initial PCI interface

DrDeano opened the pull request:
Closes #244

SamTebbs33 commented on src/kernel/arch/x86/arch.zig:
```diff
@@ -132,6 +132,37 @@ pub fn inb(port: u16) u8 {
+///     IN port: u16 - The port to write to.
+///     IN data: u32 - The byte of data that will be sent.
+///
+pub fn outl(port: u16, data: u32) void {
```
I think something like `out32` would be a better name.

DrDeano replied:
Or could have one `out` and pass a comptime type that is either u8, u16 or u32

SamTebbs33 replied:
That's an option. An alternative would be to make `data` be of type `var` and use `@TypeOf' to get the type.

DrDeano replied:
I don't think you can return a anytype. So will need to pass a comtime type. (This is for the `in` function)

SamTebbs33 replied:
You can return `@TypeOf(data)`.

DrDeano replied:
~~Ah yes :P~~

DrDeano replied:
For `in` there is no data to pass in so can't use `@TypeOf(data)` for the return

SamTebbs33 replied:
True, `in` could have a type given then.

SamTebbs33 commented on src/kernel/arch/x86/arch.zig:
```diff
@@ -132,6 +132,37 @@ pub fn inb(port: u16) u8 {
+/// Return: u32
+///     The data that the port returns.
+///
+pub fn inl(port: u16) u32 {
```
Like `out32`, I think `in32` would be better.

SamTebbs33 commented on src/kernel/pci.zig:
```diff
@@ -0,0 +1,310 @@
+    }
+
+    ///
+    /// Read the configuration register date from this device, function and register. This will
```
s/date/data

SamTebbs33 commented on src/kernel/pci.zig:
```diff
@@ -0,0 +1,310 @@
+
+    ///
+    /// Read the configuration register date from this device, function and register. This will
+    /// return the types width of the register, as PCI reads can only be done on a u32 basis.
```
I don't quite understand the role of last part of this sentence. It says that reads can only be done with u32s, which I think is a bit confusing since it returns a integer of a type depending on the register width.

DrDeano replied:
Have improved the comment

SamTebbs33 commented on src/kernel/pci.zig:
```diff
@@ -0,0 +1,310 @@
+        arch.outl(CONFIG_ADDRESS, @bitCast(u32, address) & 0xFFFFFFFC);
+        // Read the data
+        const result = arch.inl(CONFIG_DATA);
+        // Return the size the user wants
+        const shift = switch (pci_reg.getWidth()) {
+            u8 => (@intCast(u5, address.register_offset & 0x3)) * 8,
+            u16 => (@intCast(u5, address.register_offset & 0x2)) * 8,
+            u32 => 0,
+            else => @compileError("Invalid read size. Only u8, u16 and u32 allowed."),
+        };
+        comptime const mask = switch (pci_reg.getWidth()) {
+            u8 => 0xFF,
+            u16 => 0xFFFF,
+            u32 => 0xFFFFFFFF,
+            else => @compileError("Invalid read size. Only u8, u16 and u32 allowed."),
+        };
+        return @intCast(pci_reg.getWidth(), (result >> shift) & mask);
```
Can you use @truncate or some of the other builtins to do any of this work? You should be able to replace the masking with @truncate.

DrDeano replied:
Yes. I was using `@intcast` as it has runtime safety, but `@truncate` would be better.

DrDeano replied:
I still need the masking as 0x2 can't be truncated. Plus it is comptime, so there isn't any runtime cost.

SamTebbs33 replied:
Even if there's no runtime cost, it's cleaner to use what the language provides. What do you mean by 0x2 not being truncatable?

DrDeano replied:
Actually, I'm not sure how you would use `@truncate` over `@intcast`, could you give an example?

SamTebbs33 replied:
`@truncate(pci_reg.getWidth(), result >> shift)` Then you don't need the mask.

DrDeano replied:
Ah, I see, this would only be for the last `@intCast`

SamTebbs33 replied:
Yeah sorry :)

SamTebbs33 commented on src/kernel/pci.zig:
```diff
@@ -0,0 +1,310 @@
+///
+/// List all the PCI device. This is printed to the serial output.
+///
+fn listAllPCI() void {
```
I think it would be better to find all of the devices in `init` or some other helper funciton, return them and list them in kmain. That way they can be printed multiple times without having to re-find them and they can be stored and processed later on.

SamTebbs33 commented on src/kernel/kmain.zig:
```diff
@@ -169,6 +170,8 @@ fn initStage2() noreturn {
 
     tty.print("Hello Pluto from kernel :)\n", .{});
 
+    pci.init();
```
PCI is specific to x86 so we should make the device system more abstract in the future.

DrDeano replied:
I wasn't sure of this. Does ARM have PCI?

SamTebbs33 replied:
So the Arm architecture doesn't specify how devices are detected and such, that's all up to the board itself. The Raspberry Pi, which is what we're porting to at the moment, doesn't have PCI, so we'll need to make the device system abstract so each arch and board port decides how it finds the devices itself. I'm happy for this abstraction to come in another PR though.

SamTebbs33 commented on src/kernel/arch/x86/arch.zig:
```diff
@@ -102,41 +106,67 @@ pub const MEMORY_BLOCK_SIZE: usize = paging.PAGE_SIZE_4KB;
+///     IN port: u16           - The port to write to.
+///     IN data: Type          - The data that will be sent.
+///
+pub fn out(comptime Type: type, port: u16, data: Type) void {
```
We don't need the first type argument if data becomes a `var` type.

DrDeano committed:
Initial PCI interface

Closes #244

Move PCI to arch

Plus spelling

Added new out and in functions


Added new out and in to mocking


Return pci devices as a list


Improved comment


Removed mask for the return


Removed type for OUT


Added new types

SamTebbs33 reviewed (approved):
Nice one.
